kyuubi-server/src/main/scala/org/apache/kyuubi/operation/LaunchEngine.scala

```diff
@@ -38,21 +42,19 @@ class LaunchEngine(session: KyuubiSessionImpl, override val shouldRunAsync: Bool
   }
 
   override protected def afterRun(): Unit = {
+    if (shouldRunAsync) {
+      waitLaunchEngineComplete()
+    }
     OperationLog.removeCurrentOperationLog()
```

Comment:
It's safe to remove it directly from thread local to me

Reply:
De we rely on the thread local log to write anything else inside this operation after run?

Reply:
I think there is still remind operation log to write after this thread local is removed.
is it safe?

Reply:
for example, for async mode, the launch engine is still running in backend.

But it might run `afterRun` before the launch engine operation finished.

Then if the launch engine failed eventually, it may do not recored the error info into launch engine operation log.

Reply:
register and remove to thread local is just for a code path for writing. I think it's safe. If we didn't handle them in pairs, then I guess it's unsafe for potential memory leak

Reply:
but I did test for the case I mentioned.

It seems that, remove thread local directly does not matter.

Reply:
> But it might run afterRun before the launch engine operation finished.

Yes, in this case, we actually need to register the log in that async exec thread and remove it after. But this operation seems to have no log output in the async exec thread directly, so register/remove seem to be able to omitted as we do now.

register and remove in beforeRun/afterRun take affect in the FE's handler thread, it logs nothing in the async exec thread

Reply:
thanks